File: ios/Sources/Mia21/Models/ChatModels.swift
//
//  ChatModels.swift
//  Mia21
//
//  Created on November 4, 2025.
//  Copyright © 2025 Mia21. All rights reserved.
//
//  Description:
//  Chat-related data models.
//  Includes chat messages, options, responses, and tool calls.
//

import Foundation

// MARK: - Chat Message

public struct ChatMessage: Codable {
  public let role: MessageRole
  public let content: String

  public init(role: MessageRole, content: String) {
    self.role = role
    self.content = content
  }
}

// MARK: - Chat Options

public struct ChatOptions {
  public var spaceId: String?
  public var botId: String?
  public var conversationId: String?
  public var temperature: Double?
  public var maxTokens: Int?
  public var customerLlmKey: String?
  public var spaceConfig: SpaceConfig?
  public var llmType: LLMType?
  /// Voice ID for per-request voice override (ElevenLabs voice ID).
  /// Priority: Request-level voiceId > Bot-level voice_id > Default
  public var voiceId: String?

  public init(
    spaceId: String? = nil,
    botId: String? = nil,
    conversationId: String? = nil,
    temperature: Double? = nil,
    maxTokens: Int? = nil,
    customerLlmKey: String? = nil,
    spaceConfig: SpaceConfig? = nil,
    llmType: LLMType? = nil,
    voiceId: String? = nil
  ) {
    self.spaceId = spaceId
    self.botId = botId
    self.conversationId = conversationId
    self.temperature = temperature
    self.maxTokens = maxTokens
    self.customerLlmKey = customerLlmKey
    self.spaceConfig = spaceConfig
    self.llmType = llmType
    self.voiceId = voiceId
  }
}

// MARK: - Chat Response

public struct ChatResponse: Codable {
  public let message: String
  public let userId: String
  public let toolCalls: [ToolCall]?
  
  // Note: No custom CodingKeys needed - APIClient uses .convertFromSnakeCase
}

// MARK: - Tool Call

public struct ToolCall: Codable {
  public let name: String?
  public let arguments: String?
}

// MARK: - OpenAI-Compatible Completion Options

/// Options for OpenAI-compatible `/v1/chat/completions` endpoint.
/// Fully compatible with OpenAI's API - standard OpenAI SDKs work by just changing the base_url.
/// Mia21 extensions are passed via HTTP headers.
public struct CompletionOptions {
  // MARK: - OpenAI Standard Parameters (in request body)
  
  /// Model to use (e.g., "gpt-4o", "gpt-4o-mini")
  public var model: String
  /// Temperature for response randomness (0.0 - 2.0)
  public var temperature: Double?
  /// Maximum tokens in response
  public var maxTokens: Int?
  /// Whether to stream the response
  public var stream: Bool
  
  // MARK: - Mia21 Extensions (via HTTP headers)
  
  /// Space ID for memory isolation (X-Space-Id header, default: "default")
  public var spaceId: String?
  /// Agent ID for specific agent behavior (X-Agent-Id header)
  public var agentId: String?
  /// Enable voice output (X-Voice-Enabled header, default: false)
  public var voiceEnabled: Bool?
  /// ElevenLabs voice ID for TTS (X-Voice-Id header)
  public var voiceId: String?
  /// Disable memory - no history used or saved (X-Incognito header, default: false)
  public var incognito: Bool?
  
  public init(
    model: String = "gpt-4o",
    temperature: Double? = nil,
    maxTokens: Int? = nil,
    stream: Bool = false,
    spaceId: String? = nil,
    agentId: String? = nil,
    voiceEnabled: Bool? = nil,
    voiceId: String? = nil,
    incognito: Bool? = nil
  ) {
    self.model = model
    self.temperature = temperature
    self.maxTokens = maxTokens
    self.stream = stream
    self.spaceId = spaceId
    self.agentId = agentId
    self.voiceEnabled = voiceEnabled
    self.voiceId = voiceId
    self.incognito = incognito
  }
  
  /// Backward compatibility initializer with botId
  @available(*, deprecated, message: "Use agentId instead of botId")
  public init(
    spaceId: String? = nil,
    botId: String?,
    model: String = "gpt-4o",
    temperature: Double? = nil,
    maxTokens: Int? = nil,
    stream: Bool = false
  ) {
    self.model = model
    self.temperature = temperature
    self.maxTokens = maxTokens
    self.stream = stream
    self.spaceId = spaceId
    self.agentId = botId  // Map botId to agentId
    self.voiceEnabled = nil
    self.voiceId = nil
    self.incognito = nil
  }
}

// MARK: - OpenAI-Compatible Completion Response

/// Response from OpenAI-compatible `/v1/chat/completions` endpoint
public struct CompletionResponse: Codable {
  public let id: String?
  public let object: String?
  public let created: Int?
  public let model: String?
  public let choices: [CompletionChoice]?
  public let usage: CompletionUsage?
}

public struct CompletionChoice: Codable {
  public let index: Int?
  public let message: CompletionMessage?
  public let delta: CompletionMessage?
  public let finishReason: String?
}

public struct CompletionMessage: Codable {
  public let role: String?
  public let content: String?
}

public struct CompletionUsage: Codable {
  public let promptTokens: Int?
  public let completionTokens: Int?
  public let totalTokens: Int?
}

// MARK: - OpenAI-Compatible Chat Initialize (Greeting Generation)

/// Options for OpenAI-compatible `/v1/chat/initialize` endpoint.
/// Generates a personalized greeting based on user's conversation history and memories.
/// All parameters are passed via HTTP headers (no request body).
public struct GreetingOptions {
  /// Space ID for context separation (X-Space-Id header)
  public var spaceId: String?
  /// Agent ID for greeting style (X-Agent-Id header)
  public var agentId: String?
  /// Enable voice in response (X-Voice-Enabled header)
  public var voiceEnabled: Bool?
  /// Voice ID for TTS (X-Voice-Id header)
  public var voiceId: String?
  /// If true, no memory is used or saved (X-Incognito header)
  public var incognito: Bool?
  
  public init(
    spaceId: String? = nil,
    agentId: String? = nil,
    voiceEnabled: Bool? = nil,
    voiceId: String? = nil,
    incognito: Bool? = nil
  ) {
    self.spaceId = spaceId
    self.agentId = agentId
    self.voiceEnabled = voiceEnabled
    self.voiceId = voiceId
    self.incognito = incognito
  }
}

/// Backward compatibility alias
@available(*, deprecated, renamed: "GreetingOptions")
public typealias ChatInitializeOptions = GreetingOptions

/// Response from OpenAI-compatible `/v1/chat/initialize` endpoint
public struct GreetingResponse: Codable {
  /// Unique identifier for the request
  public let id: String?
  /// Object type (e.g., "chat.initialize")
  public let object: String?
  /// Unix timestamp of creation
  public let created: Int?
  /// Model used to generate the greeting
  public let model: String?
  /// Personalized greeting message based on conversation history
  public let greeting: String?
  /// Context about the user derived from history
  public let userContext: GreetingUserContext?
  /// Audio data if voice was enabled (base64 encoded)
  public let audio: String?
}

/// User context derived from conversation history
public struct GreetingUserContext: Codable {
  /// Number of previous conversations
  public let conversationCount: Int?
  /// Last interaction timestamp
  public let lastInteraction: String?
  /// Topics the user has discussed
  public let topics: [String]?
  /// Whether this is a returning user
  public let isReturningUser: Bool?
}

/// Backward compatibility aliases
@available(*, deprecated, renamed: "GreetingResponse")
public typealias ChatInitializeResponse = GreetingResponse
@available(*, deprecated, renamed: "GreetingUserContext")
public typealias ChatUserContext = GreetingUserContext

File: android/lib/src/main/kotlin/com/mia21/models/ChatModels.kt
/**
 * ChatModels.kt
 * Mia21
 *
 * Created on November 25, 2025.
 * Copyright © 2025 Mia21. All rights reserved.
 *
 * Description:
 * Chat-related data models.
 * Includes chat messages, options, responses, and tool calls.
 */

package com.mia21.models

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * Represents a chat message
 */
@Serializable
data class ChatMessage(
    val role: MessageRole,
    val content: String
)

/**
 * Options for sending chat messages
 */
data class ChatOptions(
    val spaceId: String? = null,
    val botId: String? = null,
    val conversationId: String? = null,
    val temperature: Double? = null,
    val maxTokens: Int? = null,
    val customerLlmKey: String? = null,
    val spaceConfig: SpaceConfig? = null,
    val llmType: LLMType? = null,
    /**
     * Voice ID for per-request voice override (ElevenLabs voice ID).
     * Priority: Request-level voiceId > Bot-level voice_id > Default
     */
    val voiceId: String? = null
)

/**
 * Response from a chat message
 */
@Serializable
data class ChatResponse(
    val message: String,
    @SerialName("user_id")
    val userId: String,
    @SerialName("tool_calls")
    val toolCalls: List<ToolCall>? = null
)

/**
 * Represents a tool call in a chat response
 */
@Serializable
data class ToolCall(
    val name: String? = null,
    val arguments: String? = null
)

/**
 * Options for OpenAI-compatible /v1/chat/completions endpoint.
 * Fully compatible with OpenAI's API - standard OpenAI SDKs work by just changing the base_url.
 * Mia21 extensions are passed via HTTP headers.
 */
data class CompletionOptions(
    // OpenAI Standard Parameters (in request body)
    
    /** Model to use (e.g., "gpt-4o", "gpt-4o-mini") */
    val model: String = "gpt-4o",
    /** Temperature for response randomness (0.0 - 2.0) */
    val temperature: Double? = null,
    /** Maximum tokens in response */
    val maxTokens: Int? = null,
    /** Whether to stream the response */
    val stream: Boolean = false,
    
    // Mia21 Extensions (via HTTP headers)
    
    /** Space ID for memory isolation (X-Space-Id header, default: "default") */
    val spaceId: String? = null,
    /** Agent ID for specific agent behavior (X-Agent-Id header) */
    val agentId: String? = null,
    /** Enable voice output (X-Voice-Enabled header, default: false) */
    val voiceEnabled: Boolean? = null,
    /** ElevenLabs voice ID for TTS (X-Voice-Id header) */
    val voiceId: String? = null,
    /** Disable memory - no history used or saved (X-Incognito header, default: false) */
    val incognito: Boolean? = null,
    /** Customer's own LLM API key for BYOK (X-LLM-API-Key header) */
    val llmApiKey: String? = null,
    /** User's timezone for context (X-Timezone header, e.g., "America/New_York") */
    val timezone: String? = null,
    /** User's name for personalization (X-User-Name header) */
    val userName: String? = null,
    /** Specific conversation ID (X-Conversation-Id header) */
    val conversationId: String? = null,
    /** Custom metadata as JSON string (X-Meta header) */
    val meta: String? = null
) {
    /** Backward compatibility constructor with botId */
    @Deprecated("Use agentId instead of botId", ReplaceWith("CompletionOptions(model, temperature, maxTokens, stream, spaceId, agentId = botId)"))
    constructor(
        spaceId: String?,
        botId: String?,
        model: String = "gpt-4o",
        temperature: Double? = null,
        maxTokens: Int? = null,
        stream: Boolean = false
    ) : this(
        model = model,
        temperature = temperature,
        maxTokens = maxTokens,
        stream = stream,
        spaceId = spaceId,
        agentId = botId  // Map botId to agentId
    )
}

/**
 * Response from OpenAI-compatible /v1/chat/completions endpoint
 */
@Serializable
data class CompletionResponse(
    val id: String? = null,
    val `object`: String? = null,
    val created: Int? = null,
    val model: String? = null,
    val choices: List<CompletionChoice>? = null,
    val usage: CompletionUsage? = null
)

@Serializable
data class CompletionChoice(
    val index: Int? = null,
    val message: CompletionMessage? = null,
    val delta: CompletionMessage? = null,
    @SerialName("finish_reason")
    val finishReason: String? = null
)

@Serializable
data class CompletionMessage(
    val role: String? = null,
    val content: String? = null
)

@Serializable
data class CompletionUsage(
    @SerialName("prompt_tokens")
    val promptTokens: Int? = null,
    @SerialName("completion_tokens")
    val completionTokens: Int? = null,
    @SerialName("total_tokens")
    val totalTokens: Int? = null
)

/**
 * Options for OpenAI-compatible /v1/chat/initialize endpoint.
 * Generates a personalized greeting based on user's conversation history and memories.
 * All parameters are passed via HTTP headers (no request body).
 */
data class GreetingOptions(
    /** Space ID for context separation (X-Space-Id header) */
    val spaceId: String? = null,
    /** Agent ID for greeting style (X-Agent-Id header) */
    val agentId: String? = null,
    /** Enable voice in response (X-Voice-Enabled header) */
    val voiceEnabled: Boolean? = null,
    /** Voice ID for TTS (X-Voice-Id header) */
    val voiceId: String? = null,
    /** If true, no memory is used or saved (X-Incognito header) */
    val incognito: Boolean? = null,
    /** Customer's own LLM API key for BYOK (X-LLM-API-Key header) */
    val llmApiKey: String? = null,
    /** User's timezone for time-aware greetings (X-Timezone header, e.g., "America/New_York") */
    val timezone: String? = null,
    /** User's name for personalized greetings (X-User-Name header) */
    val userName: String? = null,
    /** Specific conversation ID to continue (X-Conversation-Id header) */
    val conversationId: String? = null,
    /** Custom metadata as JSON string (X-Meta header) */
    val meta: String? = null
)

/** Backward compatibility alias */
@Deprecated("Use GreetingOptions instead", ReplaceWith("GreetingOptions"))
typealias ChatInitializeOptions = GreetingOptions

/**
 * Response from OpenAI-compatible /v1/chat/initialize endpoint
 */
@Serializable
data class GreetingResponse(
    /** Unique identifier for the request */
    val id: String? = null,
    /** Object type (e.g., "chat.initialization") */
    val `object`: String? = null,
    /** Unix timestamp of creation */
    val created: Int? = null,
    /** Model used to generate the greeting */
    val model: String? = null,
    /** Personalized greeting message based on conversation history */
    val greeting: String? = null,
    /** Whether the user has memory/history */
    @SerialName("has_memory")
    val hasMemory: Boolean? = null,
    /** Number of memories for this user */
    @SerialName("memory_count")
    val memoryCount: Int? = null,
    /** Conversation ID for continuing this conversation */
    @SerialName("conversation_id")
    val conversationId: String? = null,
    /** Context about the user derived from history (legacy) */
    @SerialName("user_context")
    val userContext: GreetingUserContext? = null,
    /** Audio data if voice was enabled (base64 encoded) */
    val audio: String? = null
)

/** Backward compatibility alias */
@Deprecated("Use GreetingResponse instead", ReplaceWith("GreetingResponse"))
typealias ChatInitializeResponse = GreetingResponse

/**
 * User context derived from conversation history
 */
@Serializable
data class GreetingUserContext(
    /** Number of previous conversations */
    @SerialName("conversation_count")
    val conversationCount: Int? = null,
    /** Last interaction timestamp */
    @SerialName("last_interaction")
    val lastInteraction: String? = null,
    /** Topics the user has discussed */
    val topics: List<String>? = null,
    /** Whether this is a returning user */
    @SerialName("is_returning_user")
    val isReturningUser: Boolean? = null
)

/** Backward compatibility alias */
@Deprecated("Use GreetingUserContext instead", ReplaceWith("GreetingUserContext"))
typealias ChatUserContext = GreetingUserContext

